src/gs-discovery-feed-content-provider.c

```diff
@@ -0,0 +1,274 @@
+
+	url_basename = g_path_get_basename (popular_background_url);
+	tile_cache_hash = g_compute_checksum_for_string (G_CHECKSUM_SHA256,
+	                                                 popular_background_url,
+	                                                 -1);
+	cache_identifier = g_strdup_printf ("%s-%s", tile_cache_hash, url_basename);
+	cache_filename = gs_utils_get_cache_filename ("eos-popular-app-thumbnails",
+	                                              cache_identifier,
+	                                              GS_UTILS_CACHE_FLAG_NONE,
+	                                              NULL);
+
+	/* Check to see if the file exists in the cache at the time we called this
+	 * function. If it does, then we can use this app in the results
+	 * since it will have a thumbnail */
+	if (g_file_test (cache_filename, G_FILE_TEST_EXISTS)) {
+		g_debug ("Hit cache for thumbnail when loading discovery feed %s: %s", popular_background_url, cache_filename);
+		return g_steal_pointer (&cache_filename);
+	}
+
+	return NULL;
+}
+
+static gboolean
+filter_for_discovery_feed_apps (GsApp *app, gpointer user_data)
+{
+	g_autofree gchar *app_thumbnail_filename = NULL;
+
+	if (gs_app_get_state (app) != AS_APP_STATE_AVAILABLE)
+		return FALSE;
+
+	/* Check to make sure that the app has an available thumbnail
+	 * otherwise, we need to skip it completely */
+	app_thumbnail_filename = get_app_thumbnail_cached_filename (app);
+
+	if (!app_thumbnail_filename)
+		return FALSE;
+
+	return TRUE;
+}
+
+static void
+search_done_cb (GObject *source,
+		GAsyncResult *res,
+		gpointer user_data)
+{
+	PendingSearch *search = user_data;
+	GsDiscoveryFeedContentProvider *self = search->provider;
+	guint i;
+	GVariantBuilder builder;
+	guint app_list_length;
+	g_autoptr(GError) error = NULL;
+	g_autoptr(GsAppList) list = NULL;
+
+	list = gs_plugin_loader_job_process_finish (self->plugin_loader, res, &error);
+	if (list == NULL) {
+		g_warning ("Error searching for Discovery Feed Apps: %s", error->message);
+		g_dbus_method_invocation_take_error (search->invocation, g_steal_pointer (&error));
+		pending_search_free (search);
+		g_application_release (g_application_get_default ());
+		return;
+	}
+
+	/* First, filter out any irrelevant apps. Then randomize the list -
+	 * this gives us a different order for each day as the random seed
+	 * for gs_app_list_randomize changes on a daily basis. Then, sort
+	 * the list, but only use the kudos as the sorting factor. This will
+	 * put "more relevant" apps closer to the top, but still with an
+	 * internally randomized order. */
+	gs_app_list_filter (list, filter_for_discovery_feed_apps, NULL);
+	gs_app_list_randomize (list);
+
+	app_list_length = gs_app_list_length (list);
+
+	/* Now that we have our indices, start building up a variant based
+	 * on those indices */
+	g_variant_builder_init (&builder, G_VARIANT_TYPE ("aa{sv}"));
+	for (i = 0; i < MIN (app_list_length, MAX_INSTALLABLE_APPS); ++i) {
```

Comment:
I still think it would be good to separate that to a cached variable

Reply:
I think the compiler optimizes such cases (and this patch is not mine either). If we want to make sure it's as efficient as possible without the compiler, then the app_list_length should have been const too, but since we're using the C90-vars-on-top thingy, I cannot do that...
But I will change the MAX thing.